File: packages/linkr-r/tests/testthat/test-paths.R
test_that("each path comes from its own variable, not from a sibling", {
  withr_vars <- c(
    LINKR_PROJECT = "/srv/proj", LINKR_SCRIPTS = "/srv/code",
    LINKR_DATASETS = "/mnt/big/data", LINKR_IDE = "/srv/home"
  )
  old <- Sys.getenv(names(withr_vars), unset = NA)
  do.call(Sys.setenv, as.list(withr_vars))
  on.exit({
    set <- old[!is.na(old)]
    if (length(set)) do.call(Sys.setenv, as.list(set))
    unset <- names(old)[is.na(old)]
    if (length(unset)) Sys.unsetenv(unset)
  })

  # The bindings are independent: datasets living on another volume must not be
  # derived from the project dir.
  expect_equal(linkr_project_dir(), "/srv/proj")
  expect_equal(linkr_scripts_dir(), "/srv/code")
  expect_equal(linkr_datasets_dir(), "/mnt/big/data")
  expect_equal(linkr_ide_dir(), "/srv/home")
})

test_that("an unset variable falls back to the working directory and warns", {
  old <- Sys.getenv("LINKR_DATASETS", unset = NA)
  Sys.unsetenv("LINKR_DATASETS")
  rm(list = ls(.linkr_warned), envir = .linkr_warned)
  on.exit(if (!is.na(old)) Sys.setenv(LINKR_DATASETS = old))

  expect_warning(path <- linkr_datasets_dir(), "not running inside a Linkr")
  expect_equal(path, getwd())
})

test_that("the fallback warns once, not on every call", {
  old <- Sys.getenv("LINKR_SCRIPTS", unset = NA)
  Sys.unsetenv("LINKR_SCRIPTS")
  rm(list = ls(.linkr_warned), envir = .linkr_warned)
  on.exit(if (!is.na(old)) Sys.setenv(LINKR_SCRIPTS = old))

  expect_warning(linkr_scripts_dir())
  expect_silent(linkr_scripts_dir())
})

test_that("an empty variable counts as unset", {
  old <- Sys.getenv("LINKR_PROJECT", unset = NA)
  Sys.setenv(LINKR_PROJECT = "")
  rm(list = ls(.linkr_warned), envir = .linkr_warned)
  on.exit(if (!is.na(old)) Sys.setenv(LINKR_PROJECT = old) else Sys.unsetenv("LINKR_PROJECT"))

  expect_warning(path <- linkr_project_dir())
  expect_equal(path, getwd())
})

File: packages/linkr-py/tests/test_paths.py
import os
from pathlib import Path

import pytest

from linkr import _paths


@pytest.fixture(autouse=True)
def _reset_warnings():
    _paths._warned.clear()


def test_each_path_comes_from_its_own_variable(monkeypatch):
    monkeypatch.setenv("LINKR_PROJECT", "/srv/proj")
    monkeypatch.setenv("LINKR_SCRIPTS", "/srv/code")
    monkeypatch.setenv("LINKR_DATASETS", "/mnt/big/data")
    monkeypatch.setenv("LINKR_IDE", "/srv/home")

    # The bindings are independent: datasets living on another volume must not be
    # derived from the project dir.
    assert _paths.project_dir() == Path("/srv/proj")
    assert _paths.scripts_dir() == Path("/srv/code")
    assert _paths.datasets_dir() == Path("/mnt/big/data")
    assert _paths.ide_dir() == Path("/srv/home")


def test_unset_variable_falls_back_to_cwd_and_warns(monkeypatch):
    monkeypatch.delenv("LINKR_DATASETS", raising=False)

    with pytest.warns(RuntimeWarning, match="not running inside a Linkr"):
        assert _paths.datasets_dir() == Path(os.getcwd())


def test_fallback_warns_once(monkeypatch, recwarn):
    monkeypatch.delenv("LINKR_SCRIPTS", raising=False)

    with pytest.warns(RuntimeWarning):
        _paths.scripts_dir()
    recwarn.clear()
    _paths.scripts_dir()
    assert len(recwarn) == 0


def test_empty_variable_counts_as_unset(monkeypatch):
    monkeypatch.setenv("LINKR_PROJECT", "")

    with pytest.warns(RuntimeWarning):
        assert _paths.project_dir() == Path(os.getcwd())
